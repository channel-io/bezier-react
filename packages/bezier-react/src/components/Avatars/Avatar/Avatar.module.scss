@layer components {
  .Avatar {
    position: relative;
    display: block;
    flex: none;

    &:where(.disabled) {
      pointer-events: none;
      opacity: var(--opacity-disabled);
    }

    &:where(.size-20) {
      width: 20px;
      height: 20px;
    }

    &:where(.size-24) {
      width: 24px;
      height: 24px;
    }

    &:where(.size-30) {
      width: 30px;
      height: 30px;
    }

    &:where(.size-36) {
      width: 36px;
      height: 36px;
    }

    &:where(.size-42) {
      width: 42px;
      height: 42px;
    }

    &:where(.size-48) {
      width: 48px;
      height: 48px;
    }

    &:where(.size-72) {
      width: 72px;
      height: 72px;
    }

    &:where(.size-90) {
      width: 90px;
      height: 90px;
    }

    &:where(.size-120) {
      width: 120px;
      height: 120px;
    }
  }

  .AvatarImage {
    --b-avatar-status-gap: -2px;
    --b-avatar-computed-status-gap: var(--b-avatar-status-gap);

    position: relative;
    box-sizing: content-box;
    display: flex;
    flex-shrink: 0;
    align-items: center;
    justify-content: center;
    width: 100%;
    height: 100%;
    outline: none;
    
    &:where(.big-size) {
      --b-avatar-status-gap: 4px;
    }

    &:where(.bordered[data-state="enabled"]) {
      --b-avatar-computed-status-gap: calc(var(--b-avatar-status-gap) + (2 * var(--b-alpha-smooth-corners-box-shadow-spread-radius)));
    }
  }

  .StatusWrapper {
    position: absolute;
    right: var(--b-avatar-computed-status-gap);
    bottom: var(--b-avatar-computed-status-gap);
    display: flex;
  }
}
