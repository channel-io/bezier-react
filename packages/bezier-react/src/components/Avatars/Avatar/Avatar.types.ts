import type {
  AlphaBezierComponentProps,
  ChildrenProps,
  DisableProps,
  SizeProps,
} from '~/src/types/ComponentProps'

import { type StatusType } from '~/src/components/Status'

export enum AvatarSize {
  Size20 = 20,
  Size24 = 24,
  Size30 = 30,
  Size36 = 36,
  Size42 = 42,
  Size48 = 48,
  Size72 = 72,
  Size90 = 90,
  Size120 = 120,
}

interface AvatarOwnProps {
  /**
   * Semantic name of the avatar.
   */
  name: string
  /**
   * Asset image URL of the avatar.
   * @default ''
   */
  avatarUrl?: string
  /**
   * Fallback image URL for the avatar.
   * @default DEFAULT_AVATAR_URL
   */
  fallbackUrl?: string
  /**
   * Additional status type for this avatar.
   * Typically, this is used to indicate the status of the user (online, offline, etc).
   *
   * If additional customization of the status component is needed,
   * pass the custom status component as `children` of the Avatar component.
   */
  status?: StatusType
  /**
   * Whether to display the border of the avatar.
   * @default false
   */
  showBorder?: boolean
  /**
   * Whether to use the smooth corners feature.
   * Overrides whether the smooth corners feature is applied when it is enabled.
   * @default true
   */
  smoothCorners?: boolean
}

export default interface AvatarProps extends
  AlphaBezierComponentProps,
  React.HTMLAttributes<HTMLDivElement>,
  SizeProps<AvatarSize>,
  DisableProps,
  ChildrenProps,
  AvatarOwnProps {}
