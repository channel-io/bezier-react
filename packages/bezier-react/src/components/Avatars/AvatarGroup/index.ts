import { AvatarGroup } from './AvatarGroup'
import { AvatarGroupEllipsisType } from './AvatarGroup.types'
import type AvatarGroupProps from './AvatarGroup.types'

export type {
  AvatarGroupProps,
}

export {
  AvatarGroup,
  AvatarGroupEllipsisType,
}
