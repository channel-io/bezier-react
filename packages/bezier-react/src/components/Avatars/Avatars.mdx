import {
  ArgsTable,
  Canvas,
  Story,
} from '@storybook/addon-docs'
import {
  Avatar,
} from 'Components/Avatars/Avatar'
import {
  AvatarGroup,
} from 'Components/Avatars/AvatarGroup'
import {
  CheckableAvatar,
} from 'Components/Avatars/CheckableAvatar'

# Avatars

## Overview

프로필 이미지를 가지는 객체를 표현하기 위해 사용합니다.

<Story id="components-avatars--overview" />

### Smoothed corners

눈썰미가 좋으신 분들은 알아보셨겠지만, **Avatar의 테두리는 완전한 원형이 아닙니다!**

CSS의 `border-radius` property를 이용하면 이러한 테두리를 얻을 수 없습니다.
[Worklet Web API](https://developer.mozilla.org/en-US/docs/Web/API/Worklet)를 사용하여, Avatar 컴포넌트만의 테두리 곡선을 직접 그려냅니다!
Worklet API를 지원하는 브라우저에서 단순한 원형 테두리가 아닌 bezier-react만의 Avatar 형태를 볼 수 있게 됩니다.

구현 방법에 대해서는 `src/worklets` 안쪽 소스 코드를 확인해보세요 :)

## Usage

### Avatar

Avatar를 사용하기 위한 최소한의 정보는 `name`과 `avatarUrl`입니다.
`name`은 Avatar에 대한 label이 됩니다. `avatarUrl`은 Avatar 컴포넌트로 표현할 이미지의 리소스 URL입니다.

<Story id="components-avatars--usage-basic" />

```tsx
<Avatar name="<name>" avatarUrl="<avatar-url>" />

// Example
<Avatar
  name="Channel.io"
  avatarUrl="https://cf.channel.io/thumb/200x200/pub-file/1/606d87d059a6093594c0/ch-symbol-filled-smiley-bg.png"
/>
```

`avatarUrl` prop으로 주어진 resource URL의 이미지를 불러오는 동안에는, fallback 이미지를 보여줍니다.
리소스 이미지를 불러오는 데 실패했을 때도 fallback 이미지를 보여줍니다.
Fallback 이미지는 bezier-react에서 기본적으로 제공하나, `fallbackUrl` prop을 통해 직접 지정하는 것도 가능합니다.

Avatar 컴포넌트에 대해 더 알아보려면 [스토리](/story/components-avatars-avatar--primary)를 살펴보세요.

### Avatar with preset status

Avatar 우측 하단에 컴포넌트를 추가하여, 아바타로 표현하고자 하는 객체의 상태를 보여줍니다.
온라인, 오프라인, 방해금지 등의 상태를 보여줄 수 있습니다.

Preset status를 사용할 때는 `status` prop에 `StatusType` enum의 값을 지정합니다.

<Story id="components-avatars--usage-preset-status" />

```tsx
<Avatar
  name="Channel.io"
  avatarUrl="https://cf.channel.io/thumb/200x200/pub-file/1/606d87d059a6093594c0/ch-symbol-filled-smiley-bg.png"
  status={StatusType.Online}
/>
```

Avatar 크기가 달라져도 preset status 컴포넌트의 크기는 변하지 않습니다.
커스텀 status 컴포넌트를 사용할 때도 유의해야 합니다.

<Story id="components-avatars--usage-preset-status-with-size" />

Preset status에 대해 더 알아보려면 [스토리](/story/components-status--primary)를 살펴보세요.

### Avatar with custom status

Preset status 이외에도, 원하는 컴포넌트를 status로 지정할 수 있습니다.
외부 서비스와의 연동 상태를 표현하거나, 이모지를 사용하여 상태를 표현할 수 있습니다.

커스텀 컴포넌트를 status로 사용할 때는, `<Avatar>` 컴포넌트의 **children**으로 지정합니다.

<Story id="components-avatars--usage-custom-status" />

```tsx
<Avatar
  name="Channel.io"
  avatarUrl="https://cf.channel.io/thumb/200x200/pub-file/1/606d87d059a6093594c0/ch-symbol-filled-smiley-bg.png"
>
  <Emoji // <- status 자리에 들어갈 컴포넌트를 children으로 지정합니다.
    name="smile"
    avatarUrl="..."
  />
</Avatar>
```

다른 Avatar 컴포넌트를 status로 사용할 때는, `showBorder` prop을 `true`로 지정하여 원본 Avatar와의 대비를 주어야 합니다.

<Story id="components-avatars--usage-custom-status-avatar" />

```tsx
<Avatar
  name="Channel.io"
  avatarUrl="https://cf.channel.io/thumb/200x200/pub-file/1/606d87d059a6093594c0/ch-symbol-filled-smiley-bg.png"
>
  <Avatar
    name="status"
    avatarUrl="https://bit.ly/code-beast"
    showBorder // <-- showBorder prop을 지정하면, 원본 Avatar와의 대비가 보입니다.
  />
</Avatar>
```

### Disabled avatar

`disabled` prop을 사용하여, 비활성화 상태의 Avatar를 표현합니다.

<Story id="components-avatars--usage-disabled" />

```tsx
<Avatar
  name="Channel.io"
  avatarUrl="https://cf.channel.io/thumb/200x200/pub-file/1/606d87d059a6093594c0/ch-symbol-filled-smiley-bg.png"
  disabled
/>
```

### Checkable avatar

Avatar 컴포넌트를 통해 표현되는 객체들이 있을 때, 객체들의 목록에서 선택하는 UX를 표현하기 위해 checkable avatar를 사용합니다.

Checkable avatar는 선택 여부를 toggle할 수 있으며, 선택한 상태에서는 이미지 대신 체크 아이콘이 표시됩니다.

<Story id="components-avatars--usage-checkable-avatar" />

Avatar 컴포넌트와 유사하지만, `isChecked` 와 `isCheckable` prop을 통해 체크 상태를 표현할 수 있는 `CheckableAvatar` 컴포넌트를 사용하세요.

```tsx
<CheckableAvatar
  name="Channel.io"
  avatarUrl="https://cf.channel.io/thumb/200x200/pub-file/1/606d87d059a6093594c0/ch-symbol-filled-smiley-bg.png"
  isChecked
/>
```

Checkable avatar에 대해 더 알아보려면 [스토리](/story/components-avatars-checkableavatar--primary)를 살펴보세요.

### Multiple avatars (AvatarGroup)

Avatar 목록을 표현하기 위해서는 AvatarGroup 컴포넌트를 사용합니다.

목록에 펼쳐놓을 최대 개수를 `max` prop으로 지정하고, 이보다 긴 목록은 두 가지 ellipsis 방식을 이용하여 표현합니다.
Ellipsis 방법은 `ellipsisType` prop을 통해 지정합니다.
AvatarGroup의 children으로 목록으로 표현할 Avatar 컴포넌트를 전달합니다.

<Story id="components-avatars--usage-group-ellipsis" />

```tsx
<AvatarGroup
  max={4}
  ellipsisType={AvatarGroupEllipsisType.Icon}
>
  <Avatar ... />
  <Avatar ... />
  <Avatar ... />
  <Avatar ... />
  <Avatar ... />
  <Avatar ... />
</AvatarGroup>


<AvatarGroup
  max={4}
  ellipsisType={AvatarGroupEllipsisType.Count}
>
  <Avatar ... />
  <Avatar ... />
  <Avatar ... />
  <Avatar ... />
  <Avatar ... />
  <Avatar ... />
</AvatarGroup>
```

AvatarGroup으로 표현하는 Avatar 목록은 같은 크기나 border property를 가져야 하기 때문에,
Avatar 각각에 property를 지정하기보다는 AvatarGroup 컴포넌트에 한 번에 property를 지정하는 것이 best practice입니다.

```tsx
// DON'T ❌
<AvatarGroup ... >
  <Avatar size={AvatarSize.Size24} ... />
  <Avatar size={AvatarSize.Size24} ... />
  <Avatar size={AvatarSize.Size24} ... />
</AvatarGroup>

// DO ✅
<AvatarGroup size={AvatarSize.Size24} ... >
  <Avatar ... />
  <Avatar ... />
  <Avatar ... />
</AvatarGroup>
```

Group 내 Avatar 간의 간격을 조절하기 위해서는 `spacing` prop을 사용하세요.
`spacing`이 음수가 되는 경우 (즉, Avatar 간에 겹치는 경우) 또한 종종 사용되는 UX입니다.

<Story id="components-avatars--usage-group-spacing" />

Avatar group에 대해 더 알아보려면 [스토리](/story/components-avatars-avatargroup--primary)를 살펴보세요.

## Variants

### Size

20x20 부터 120x120 까지, preset size가 존재합니다.
기본 크기는 24x24 입니다.

<Story id="components-avatars--variants-size" />

## API

### Avatar

<ArgsTable of={Avatar} />

### AvatarGroup

<ArgsTable of={AvatarGroup} />

### CheckableAvatar

<ArgsTable of={CheckableAvatar} />

## Version

- Available since 0.3.21
