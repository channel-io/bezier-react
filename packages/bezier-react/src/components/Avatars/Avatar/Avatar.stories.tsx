/* External dependencies */
import React from 'react'
import { base } from 'paths.macro'
import { Story, Meta } from '@storybook/react'

/* Internal dependencies */
import { styled } from 'Foundation'
import { getTitle } from 'Utils/storyUtils'
import { StatusType } from 'Components/Status'
import { Avatar } from './Avatar'
import AvatarProps, { AvatarSize } from './Avatar.types'

const MOCK_AVATAR_URL = 'https://cf.channel.io/thumb/200x200/pub-file/1/606d87d059a6093594c0/ch-symbol-filled-smiley-bg.png'

const avatarSizeList = Object.keys(AvatarSize)
  .filter(value => Number.isNaN(Number(value)) === true)
  .map(key => AvatarSize[key])

const statusTypeList = Object.keys(StatusType)
  .map(key => StatusType[key])

export default {
  title: getTitle(base),
  component: Avatar,
  argTypes: {
    size: {
      control: {
        type: 'radio',
        options: avatarSizeList,
      },
    },
    status: {
      control: {
        type: 'radio',
        options: [
          undefined,
          ...statusTypeList,
        ],
      },
    },
    onClick: {
      action: 'clicked',
    },
    onMouseEnter: {
      action: 'mouseEnter',
    },
    onMouseLeave: {
      action: 'mouseLeave',
    },
  },
} as Meta

// NOTE: (@ed) border 색상을 명확하게 보여주기 위해 회색의 Wrapper를 추가했습니다
const Wrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 200px;
  height: 200px;
  background-color: ${({ foundation }) => foundation?.theme?.['bg-grey-light']};
`

const Template: Story<AvatarProps> = (args) => (
  <Wrapper>
    <Avatar {...args} />
  </Wrapper>
)

export const Primary: Story<AvatarProps> = Template.bind({})
Primary.args = {
  avatarUrl: MOCK_AVATAR_URL,
  name: 'Channel',
  size: AvatarSize.Size24,
  showBorder: false,
  disabled: false,
}

const TemplateWithCustomStatus: Story<AvatarProps> = (args) => (
  <Wrapper>
    <Avatar {...args}>
      <Avatar
        avatarUrl="https://bit.ly/kent-c-dodds"
        name="Kent Dodds"
        size={AvatarSize.Size20}
        showBorder
      />
    </Avatar>
  </Wrapper>
)

export const WithCustomStatus: Story<AvatarProps> = TemplateWithCustomStatus.bind({})
WithCustomStatus.args = {
  avatarUrl: MOCK_AVATAR_URL,
  name: 'Channel',
  size: AvatarSize.Size48,
  showBorder: false,
  disabled: false,
}
