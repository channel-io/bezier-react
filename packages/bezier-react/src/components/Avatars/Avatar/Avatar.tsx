import React, {
  forwardRef,
  useMemo,
} from 'react'

import classNames from 'classnames'

import { isEmpty } from '~/src/utils/typeUtils'

import { type BoxShadow } from '~/src/components/AlphaSmoothCornersBox'
import { AVATAR_BORDER_RADIUS_PERCENTAGE } from '~/src/components/Avatars/AvatarStyle'
import {
  Status,
  StatusSize,
} from '~/src/components/Status'

// NOTE: Don't fix it. When using absolute paths, a module resolve error occurs at the runtime of the library consumer.
// eslint-disable-next-line no-restricted-imports
import defaultAvatarUrl from '../assets/defaultAvatar.svg'

import type AvatarProps from './Avatar.types'
import { AvatarSize } from './Avatar.types'
import useProgressiveImage from './useProgressiveImage'

import * as Styled from './Avatar.styled'

const shadow: BoxShadow = {
  spreadRadius: 2,
  color: 'bg-white-high',
}

export const AVATAR_WRAPPER_TEST_ID = 'bezier-react-avatar-wrapper'
export const AVATAR_TEST_ID = 'bezier-react-avatar'
export const STATUS_WRAPPER_TEST_ID = 'bezier-react-status-wrapper'

export const Avatar = forwardRef(function Avatar({
  avatarUrl = '',
  fallbackUrl = defaultAvatarUrl,
  size = AvatarSize.Size24,
  name,
  testId = AVATAR_TEST_ID,
  disabled = false,
  showBorder = false,
  smoothCorners = true,
  status,
  className: classNameProp,
  wrapperClassName,
  interpolation,
  wrapperInterpolation,
  style: styleProp,
  children,
  ...rest
}: AvatarProps, forwardedRef: React.Ref<HTMLDivElement>) {
  const loadedAvatarUrl = useProgressiveImage(avatarUrl, fallbackUrl)

  const StatusComponent = useMemo(() => {
    if (
      (isEmpty(children) && !status)
      || (children && !React.isValidElement(children))
    ) {
      return null
    }

    const statusSize = size >= AvatarSize.Size90 ? StatusSize.L : StatusSize.M

    const Contents = (() => {
      if (children) { return children }
      if (status) {
        return (
          <Status
            type={status}
            size={statusSize}
          />
        )
      }
      return null
    })()

    return (
      <Styled.StatusWrapper data-testid={STATUS_WRAPPER_TEST_ID}>
        { Contents }
      </Styled.StatusWrapper>
    )
  }, [
    status,
    size,
    children,
  ])

  const style = useMemo(() => ({
    ...styleProp,
    '--bezier-avatar-status-gap': `${size >= AvatarSize.Size72 ? 4 : -2}px`,
  }), [
    styleProp,
    size,
  ])

  const className = classNames(
    classNameProp,
    size && `size-${size}`,
    showBorder && 'bordered',
  )

  return (
    <Styled.AvatarWrapper
      data-testid={AVATAR_WRAPPER_TEST_ID}
      data-disabled={disabled}
      className={wrapperClassName}
      interpolation={wrapperInterpolation}
    >
      <Styled.AvatarImage
        {...rest}
        ref={forwardedRef}
        data-testid={testId}
        aria-label={name}
        style={style}
        className={className}
        interpolation={interpolation}
        disabled={!smoothCorners}
        borderRadius={`${AVATAR_BORDER_RADIUS_PERCENTAGE}%`}
        shadow={showBorder ? shadow : undefined}
        backgroundColor="bg-white-normal"
        backgroundImage={loadedAvatarUrl}
      >
        { StatusComponent }
      </Styled.AvatarImage>
    </Styled.AvatarWrapper>
  )
})
