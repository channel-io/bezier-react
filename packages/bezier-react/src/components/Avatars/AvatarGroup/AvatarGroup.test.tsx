/* External dependencies */
import React from 'react'

/* Internal dependencies */
import { render } from '~/src/utils/testUtils'
import { Avatar } from '~/src/components/Avatars/Avatar'
import MOCK_AVATAR_LIST from './__mocks__/avatarList'
import {
  AvatarGroup,
  AVATAR_GROUP_ELLIPSIS_ICON_TEST_ID,
} from './AvatarGroup'
import {
  AvatarGroupEllipsisType,
  type AvatarGroupProps,
} from './AvatarGroup.types'

jest.mock('~/src/worklets/EnableCSSHoudini', () => ({
  __esModule: true,
  ...jest.requireActual('~/src/worklets/EnableCSSHoudini') as object,
  enableSmoothCorners: { current: true },
}))

describe('AvatarGroup', () => {
  let props: AvatarGroupProps
  const mockFallbackUrl = 'https://www.google.com'

  beforeEach(() => {
    props = {
      max: MOCK_AVATAR_LIST.length - 1,
      spacing: 4,
      ellipsisType: AvatarGroupEllipsisType.Icon,
    }
  })

  afterAll(() => {
    jest.restoreAllMocks()
  })

  const renderComponent = (otherProps?: AvatarGroupProps) => render(
    <AvatarGroup {...props} {...otherProps}>
      { MOCK_AVATAR_LIST.map(({ id, avatarUrl, name }) => (
        <Avatar
          key={id}
          avatarUrl={avatarUrl}
          fallbackUrl={mockFallbackUrl}
          name={name}
        />
      )) }
    </AvatarGroup>,
  )

  describe('Ellipsis type - Icon', () => {
    beforeEach(() => {
      props.ellipsisType = AvatarGroupEllipsisType.Icon
    })

    it('Snapshot', () => {
      const { getByRole } = renderComponent()
      const rendered = getByRole('group')
      expect(rendered).toMatchSnapshot()
    })

    it('should render ellipsis icon when avatar count is more than max', () => {
      const { getByTestId } = renderComponent()
      const rendered = getByTestId(AVATAR_GROUP_ELLIPSIS_ICON_TEST_ID)
      expect(rendered).toBeInTheDocument()
    })

    it('should not render ellipsis icon when avatar count is less than max', () => {
      props.max = MOCK_AVATAR_LIST.length
      const { queryByTestId } = renderComponent()
      const rendered = queryByTestId(AVATAR_GROUP_ELLIPSIS_ICON_TEST_ID)
      expect(rendered).not.toBeInTheDocument()
    })
  })

  describe('Ellipsis type - Count', () => {
    beforeEach(() => {
      props.ellipsisType = AvatarGroupEllipsisType.Count
    })

    it('Snapshot', () => {
      const { getByRole } = renderComponent()
      const rendered = getByRole('group')
      expect(rendered).toMatchSnapshot()
    })

    it('should render ellipsis count when avatar count is more than max', () => {
      const { getByText } = renderComponent()
      const rendered = getByText('+1')
      expect(rendered).toBeInTheDocument()
    })

    it('should not render ellipsis count when avatar count is less than max', () => {
      props.max = MOCK_AVATAR_LIST.length
      const { queryByText } = renderComponent()
      const rendered = queryByText('+1')
      expect(rendered).not.toBeInTheDocument()
    })
  })
})
