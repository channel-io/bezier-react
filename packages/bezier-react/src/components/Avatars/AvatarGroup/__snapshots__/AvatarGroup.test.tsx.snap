// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarGroup Ellipsis type - Count Snapshot 1`] = `
.c2 {
  position: relative;
  box-sizing: content-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 24px;
  height: 24px;
  outline: none;
  margin: 0px;
  background-color: #FFFFFF;
  background-image: var(--background-image);
  background-size: cover;
  border-radius: 42%;
}

.c1 {
  position: relative;
  z-index: 0;
}

.c4 {
  font-size: 1.3rem;
  line-height: 1.8rem;
  margin: 0px 0px 0px 0px;
  font-style: normal;
  font-weight: normal;
  color: inherit;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: color;
  transition-property: color;
}

.c5 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: var(--bezier-avatar-group-size);
  color: var(--txt-black-dark);
}

.c0 {
  position: relative;
  z-index: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 > * + * {
  margin-left: var(--bezier-avatar-group-spacing);
}

.c3 {
  padding-right: var(--bezier-avatar-group-ellipsis-mr);
}

.c3.c3 {
  margin-left: var(--bezier-avatar-group-ellipsis-ml);
}

@supports (background:paint(smooth-corners)) {
  .c2 {
    padding: 0px;
    margin: 0px;
    background: paint(smooth-corners);
    border-radius: 0;
    box-shadow: none;
    border-image-source: var(--background-image);
    --smooth-corners: 42%;
    --smooth-corners-shadow: 0 0 0 0 transparent;
    --smooth-corners-bg-color: #FFFFFF;
    --smooth-corners-padding: 0;
    --smooth-corners-radius-unit: string;
  }
}

<div
  class="c0"
  role="group"
  style="--bezier-avatar-group-spacing: 4px; --bezier-avatar-group-size: 24px;"
>
  <div
    class="c1"
    data-testid="bezier-react-avatar-wrapper"
  >
    <div
      aria-label="Christian Nwamba"
      class="c2"
      data-testid="bezier-react-avatar"
      role="img"
      size="24"
      style="--background-image: url(https://www.google.com);"
    />
  </div>
  <div
    class="c1"
    data-testid="bezier-react-avatar-wrapper"
  >
    <div
      aria-label="Kola Tioluwani"
      class="c2"
      data-testid="bezier-react-avatar"
      role="img"
      size="24"
      style="--background-image: url(https://www.google.com);"
    />
  </div>
  <div
    class="c1"
    data-testid="bezier-react-avatar-wrapper"
  >
    <div
      aria-label="Kent Dodds"
      class="c2"
      data-testid="bezier-react-avatar"
      role="img"
      size="24"
      style="--background-image: url(https://www.google.com);"
    />
  </div>
  <div
    class="c1"
    data-testid="bezier-react-avatar-wrapper"
  >
    <div
      aria-label="Ryan Florence"
      class="c2"
      data-testid="bezier-react-avatar"
      role="img"
      size="24"
      style="--background-image: url(https://www.google.com);"
    />
  </div>
  <div
    class="c1"
    data-testid="bezier-react-avatar-wrapper"
  >
    <div
      aria-label="Dan Abrahmov"
      class="c2"
      data-testid="bezier-react-avatar"
      role="img"
      size="24"
      style="--background-image: url(https://www.google.com);"
    />
  </div>
  <div
    class="c1"
    data-testid="bezier-react-avatar-wrapper"
  >
    <div
      aria-label="Prosper Otemuyiwa"
      class="c2"
      data-testid="bezier-react-avatar"
      role="img"
      size="24"
      style="--background-image: url(https://www.google.com);"
    />
  </div>
  <div
    class="c3"
    style="--bezier-avatar-group-ellipsis-mr: 5px; --bezier-avatar-group-ellipsis-ml: 4px;"
  >
    <span
      class="c4 c5"
      data-testid="bezier-react-text"
    >
      +1
    </span>
  </div>
</div>
`;

exports[`AvatarGroup Ellipsis type - Icon Snapshot 1`] = `
.c5 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  margin: 0px 0px 0px 0px;
  color: #FFFFFF;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: color;
  transition-property: color;
}

.c2 {
  position: relative;
  box-sizing: content-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 24px;
  height: 24px;
  outline: none;
  margin: 0px;
  background-color: #FFFFFF;
  background-image: var(--background-image);
  background-size: cover;
  border-radius: 42%;
}

.c1 {
  position: relative;
  z-index: 0;
}

.c0 {
  position: relative;
  z-index: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 > * + * {
  margin-left: var(--bezier-avatar-group-spacing);
}

.c3 {
  position: relative;
}

.c4 {
  position: absolute;
  top: 0;
  right: 0;
  z-index: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
  height: 100%;
  outline: none;
  margin: 0px;
  background-color: #00000033;
  border-radius: 42%;
}

@supports (background:paint(smooth-corners)) {
  .c2 {
    padding: 0px;
    margin: 0px;
    background: paint(smooth-corners);
    border-radius: 0;
    box-shadow: none;
    border-image-source: var(--background-image);
    --smooth-corners: 42%;
    --smooth-corners-shadow: 0 0 0 0 transparent;
    --smooth-corners-bg-color: #FFFFFF;
    --smooth-corners-padding: 0;
    --smooth-corners-radius-unit: string;
  }
}

@supports (background:paint(smooth-corners)) {
  .c4 {
    padding: 0px;
    margin: 0px;
    background: paint(smooth-corners);
    border-radius: 0;
    box-shadow: none;
    --smooth-corners: 42%;
    --smooth-corners-shadow: 0 0 0 0 transparent;
    --smooth-corners-bg-color: #00000033;
    --smooth-corners-padding: 0;
    --smooth-corners-radius-unit: string;
  }
}

<div
  class="c0"
  role="group"
  style="--bezier-avatar-group-spacing: 4px; --bezier-avatar-group-size: 24px;"
>
  <div
    class="c1"
    data-testid="bezier-react-avatar-wrapper"
  >
    <div
      aria-label="Christian Nwamba"
      class="c2"
      data-testid="bezier-react-avatar"
      role="img"
      size="24"
      style="--background-image: url(https://www.google.com);"
    />
  </div>
  <div
    class="c1"
    data-testid="bezier-react-avatar-wrapper"
  >
    <div
      aria-label="Kola Tioluwani"
      class="c2"
      data-testid="bezier-react-avatar"
      role="img"
      size="24"
      style="--background-image: url(https://www.google.com);"
    />
  </div>
  <div
    class="c1"
    data-testid="bezier-react-avatar-wrapper"
  >
    <div
      aria-label="Kent Dodds"
      class="c2"
      data-testid="bezier-react-avatar"
      role="img"
      size="24"
      style="--background-image: url(https://www.google.com);"
    />
  </div>
  <div
    class="c1"
    data-testid="bezier-react-avatar-wrapper"
  >
    <div
      aria-label="Ryan Florence"
      class="c2"
      data-testid="bezier-react-avatar"
      role="img"
      size="24"
      style="--background-image: url(https://www.google.com);"
    />
  </div>
  <div
    class="c1"
    data-testid="bezier-react-avatar-wrapper"
  >
    <div
      aria-label="Dan Abrahmov"
      class="c2"
      data-testid="bezier-react-avatar"
      role="img"
      size="24"
      style="--background-image: url(https://www.google.com);"
    />
  </div>
  <div
    class="c3"
    data-testid="bezier-react-avatar-group-ellipsis-icon"
  >
    <div
      class="c4"
    >
      <svg
        class="c5"
        color="bgtxt-absolute-white-dark"
        data-testid="bezier-react-icon"
        fill="none"
        foundation="[object Object]"
        height="16"
        marginbottom="0"
        marginleft="0"
        marginright="0"
        margintop="0"
        viewBox="0 0 24 24"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M2.5 12a2.257 2.257 0 0 0 2.25 2.25A2.257 2.257 0 0 0 7 12a2.257 2.257 0 0 0-2.25-2.25A2.257 2.257 0 0 0 2.5 12Zm9.5 2.25A2.257 2.257 0 0 1 9.75 12 2.257 2.257 0 0 1 12 9.75 2.257 2.257 0 0 1 14.25 12 2.257 2.257 0 0 1 12 14.25Zm7.25 0A2.257 2.257 0 0 1 17 12a2.257 2.257 0 0 1 2.25-2.25A2.257 2.257 0 0 1 21.5 12a2.257 2.257 0 0 1-2.25 2.25Z"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
    </div>
    <div
      class="c1"
      data-testid="bezier-react-avatar-wrapper"
    >
      <div
        aria-label="Prosper Otemuyiwa"
        class="c2"
        data-testid="bezier-react-avatar"
        role="img"
        size="24"
        style="--background-image: url(https://www.google.com);"
      />
    </div>
  </div>
</div>
`;
